spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -387,6 +391,68 @@ public void stop(Runnable callback) {
 		}
 		final KafkaMessageDrivenChannelAdapter<?, ?> kafkaMessageDrivenChannelAdapter =
 				new KafkaMessageDrivenChannelAdapter<>(messageListenerContainer);
+		kafkaMessageDrivenChannelAdapter.setMessageConverter(getMessageConverter(extendedConsumerProperties));
+		kafkaMessageDrivenChannelAdapter.setBeanFactory(this.getBeanFactory());
+		ErrorInfrastructure errorInfrastructure = registerErrorInfrastructure(destination, consumerGroup,
+				extendedConsumerProperties);
+		if (extendedConsumerProperties.getMaxAttempts() > 1) {
+			kafkaMessageDrivenChannelAdapter.setRetryTemplate(buildRetryTemplate(extendedConsumerProperties));
+			kafkaMessageDrivenChannelAdapter.setRecoveryCallback(errorInfrastructure.getRecoverer());
+		}
+		else {
+			kafkaMessageDrivenChannelAdapter.setErrorChannel(errorInfrastructure.getErrorChannel());
+		}
```

Comment:
Gary, can you please raise an "improvement" issue in core to revisit the above code? It is almost identical in Rabbit binder and in new SI-backed binder, so it seems to be begging for some abstraction and/or utility. So I'd like to revisit it at some point in the future.

Reply:
The problem is that the abstract class doesn't know anything about the channel adapter and wouldn't know how to inject the error channel. This is related to the complete overhaul that Vinny wanted to do in the fall but found that it was a huge change. Perhaps this is something we can look at making the binders/bindings more spring-friendly.